src/wallet/db.cpp

```diff
@@ -29,27 +30,24 @@ namespace {
 //! (https://docs.oracle.com/cd/E17275_01/html/programmer_reference/program_copy.html),
 //! so bitcoin should never create different databases with the same fileid, but
 //! this error can be triggered if users manually copy database files.
-void CheckUniqueFileid(const BerkeleyEnvironment& env, const std::string& filename, Db& db)
+void CheckUniqueFileid(BerkeleyEnvironment& env, const std::string& filename, Db& db)
 {
     if (env.IsMock()) return;
 
-    u_int8_t fileid[DB_FILE_ID_LEN];
-    int ret = db.get_mpf()->get_fileid(fileid);
+    BerkeleyFileid fileid;
+    int ret = db.get_mpf()->get_fileid(fileid.value);
     if (ret != 0) {
         throw std::runtime_error(strprintf("BerkeleyBatch: Can't open database %s (get_fileid failed with %d)", filename, ret));
     }
 
-    for (const auto& item : env.mapDb) {
-        u_int8_t item_fileid[DB_FILE_ID_LEN];
-        if (item.second && item.second->get_mpf()->get_fileid(item_fileid) == 0 &&
-            memcmp(fileid, item_fileid, sizeof(fileid)) == 0) {
-            const char* item_filename = nullptr;
-            item.second->get_dbname(&item_filename, nullptr);
+    for (const auto& item : env.m_fileids) {
+        if (item.first != filename && memcmp(fileid.value, item.second.value, sizeof(fileid)) == 0) {
             throw std::runtime_error(strprintf("BerkeleyBatch: Can't open database %s (duplicates fileid %s from %s)", filename,
-                HexStr(std::begin(item_fileid), std::end(item_fileid)),
-                item_filename ? item_filename : "(unknown database)"));
+                HexStr(std::begin(item.second.value), std::end(item.second.value)), item.first));
         }
     }
+
+    env.m_fileids.emplace(filename, fileid);
```

Comment:
Since `CheckUniqueFileid` is called on all `BerkeleyEnvironment` instances, this will incorrectly add the filename to unrelated BerkeleyEnvironment instances that don't actually contain `filename`. I'd suggest dropping this line, adding a `BerkeleyFileid& fileid` output parameter to `CheckUniqueFileid`, and passing in `this->env->m_fileids[strFilename]` where `CheckUniqueFileid` is called in `BerkeleyBatch::BerkeleyBatch`.

Reply:
done